### PR TITLE
Add semantic segmentation task

### PR DESCRIPTION
**Pull request by edgarriba**
## What does this PR do?

This PR implements semantic segmentation task. It includes the following data structures:
- `SemantincSegmentationPreprocess`
- `SemantincSegmentationData`
- `SemantincSegmentation`

some results:

image

**TODO:**
 - [x] tests
 - [x] docs
 - [x] example: https://colab.research.google.com/drive/12GdbLf4mB_jni1vNICDwXxNyigYKd1Y4?usp=sharing 
 - [ ] benchmarks
   - https://www.kaggle.com/nbuhagiar/carla-semantic-segmentation

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/vision/segmentation/model.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    def forward(self, x) -> torch.Tensor:
+        return self.backbone(x)['out']  # TODO: find a proper way to get 'out' from registry
+
+
+@SemanticSegmentation.backbones(name="torchvision/fcn_resnet50")
+def fn(pretrained: bool, num_classes: int):
+    import torchvision
+    model: nn.Module = torchvision.models.segmentation.fcn_resnet50(pretrained=pretrained, num_classes=num_classes)
+    return model
```
Comment (justusschock) [inline flash/vision/segmentation/model.py]: maybe only register this if torchvision is installed?

File: flash/vision/segmentation/data.py [hunk 2]
```diff
@@ -18,38 +31,14 @@
 from flash.data.process import Preprocess
 from flash.utils.imports import _KORNIA_AVAILABLE, _MATPLOTLIB_AVAILABLE
 
+from . import transforms as T
```
Comment (justusschock) [inline flash/vision/segmentation/data.py:34]: Can we stay consistent with avoiding relative imports here?

File: flash/vision/segmentation/transforms.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        self.keys = keys
+
+    @torch.no_grad()
+    def forward(self, x: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
+        for seq in self.children():
+            for aug in seq:
+                for key in self.keys:
+                    # check wether the transform was applied
+                    # and apply transform
+                    if hasattr(aug, "_params") and bool(aug._params):
```
Comment (ethanwharris) [inline flash/vision/segmentation/transforms.py:34]: Can you explain what this is doing?

Comment (edgarriba) [reply flash/vision/segmentation/transforms.py:34]: in kornia every time a forward pass is called for an augmentation - the generated random params are stored in `_params` in order to assure deterministic transformations. We reuse those parameters to be applied across the different key values.

Comment (ethanwharris) [reply flash/vision/segmentation/transforms.py:34]: I see, makes sense :smiley:

File: flash/vision/segmentation/backbones.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+
+@SEMANTIC_SEGMENTATION_BACKBONES(name="torchvision/fcn_resnet50")
+def load_torchvision_fcn_resnet50(pretrained: bool, num_classes: int) -> nn.Module:
+    model = torchvision.models.segmentation.fcn_resnet50(pretrained=pretrained)
+    model.classifier[-1] = nn.Conv2d(512, num_classes, kernel_size=(1, 1), stride=(1, 1))
+    return model
+
+
+@SEMANTIC_SEGMENTATION_BACKBONES(name="torchvision/fcn_resnet101")
```
Comment (tchaton) [inline flash/vision/segmentation/backbones.py:32]: Any library should we integrate there ? Like IceVision ?

File: flash/vision/segmentation/data.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+            predict_transform = T.default_val_transforms(image_size)
+
+        return (
+            train_transform,
+            val_transform,
+            test_transform,
+            predict_transform,
+        )
+
+    def load_sample(self, sample: Union[str, Tuple[str,
```
Comment (tchaton) [inline flash/vision/segmentation/data.py:110]: formatting is off.

File: flash/vision/segmentation/data.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+        return {SegmentationKeys.IMAGES: img, SegmentationKeys.MASKS: img_labels}
+
+    # TODO: this routine should be moved to `per_batch_transform` once we have a way to
+    # forward the labels to the loss function:.
+    def post_tensor_transform(
+        self, sample: Union[torch.Tensor, Dict[SegmentationKeys, torch.Tensor]]
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
+        if isinstance(sample, torch.Tensor):  # case for predict
+            out = sample.float() / 255.  # TODO: define predict transforms
```
Comment (tchaton) [inline flash/vision/segmentation/data.py:138]: that's a transform. It shouldn't be hard-coded.

File: flash/vision/segmentation/backbones.py (hunk 4 above)
Comment (kaushikb11) [reply flash/vision/segmentation/backbones.py:32]: Torchvision should be good enough for now.
We already have heavy dependencies.

File: flash/vision/segmentation/backbones.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+from flash.utils.imports import _TORCHVISION_AVAILABLE
+
+if _TORCHVISION_AVAILABLE:
+    import torchvision
+
+SEMANTIC_SEGMENTATION_BACKBONES = FlashRegistry("backbones")
+
+
+@SEMANTIC_SEGMENTATION_BACKBONES(name="torchvision/fcn_resnet50")
+def load_torchvision_fcn_resnet50(pretrained: bool, num_classes: int) -> nn.Module:
```
Comment (kaushikb11) [inline flash/vision/segmentation/backbones.py:26]:
```suggestion
def load_torchvision_fcn_resnet50(pretrained: bool = True, num_classes: int) -> nn.Module:
```
Good to have the defaults in hints.

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Review by tchaton — COMMENTED** [review]
Overall looks good ! Small nits

File: flash/vision/segmentation/data.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        if not isinstance(input_data, list) and not isinstance(target_data, list):
+            input_data = [input_data]
+            target_data = [target_data]
+
+        data = filter(
+            lambda sample: (
+                has_file_allowed_extension(sample[0], self.extensions) and
+                has_file_allowed_extension(sample[1], self.extensions)
+            ),
+            zip(input_data, target_data),
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply flash/vision/segmentation/data.py:75]: agree with @ethanwharris comment

File: flash/vision/segmentation/data.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+        """
+        self.image_size = image_size
+
+        super().__init__(
+            train_transform=train_transform,
+            val_transform=val_transform,
+            test_transform=test_transform,
+            predict_transform=predict_transform,
+            data_sources={DefaultDataSources.PATHS: SemanticSegmentationPathsDataSource()},
+            default_data_source=DefaultDataSources.PATHS,
```
Comment (edgarriba) [inline flash/vision/segmentation/data.py:126]: @ethanwharris cannot this be overwritten by `data_sources` key ?

File: flash/vision/segmentation/transforms.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+import kornia as K
+import torch
+import torch.nn as nn
+
+from flash.data.data_source import DefaultDataKeys
+from flash.data.transforms import ApplyToKeys
+from flash.data.utils import convert_to_modules
+
+
+class KorniaParallelTransforms(nn.Sequential):
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply flash/vision/segmentation/transforms.py:25]: @tchaton I'll port that one to kornia later - we actually got some feedback requesting this

File: flash/vision/segmentation/data.py [hunk 11]
```diff
@@ -62,10 +62,20 @@ def load_data(self, data: Union[Tuple[str, str], Tuple[List[str], List[str]]]) -
...
+        else:
+            if not isinstance(input_data, list) and not isinstance(target_data, list):
+                input_data = [input_data]
+                target_data = [target_data]
 
-        if not isinstance(input_data, list) and not isinstance(target_data, list):
-            input_data = [input_data]
-            target_data = [target_data]
+            if len(input_data) != len(target_data):
+                rank_zero_warn(
```
Comment (tchaton) [inline flash/vision/segmentation/data.py:71]: Personally, I think we should break or check the filename matches.

**Review by tchaton — APPROVED** [review]
LGTM !

**Pull request merged**